Optimize apiserver interactions for health check

tgraf opened the pull request:
The status checker was triggering a version call to the apiserver every 5 seconds. Optimize this using the following approach:

 * Use a cluster size dependent interval
 * Increase the interval while failing to ensure fast recovery
 * Consider the timestamp of the last event received to avoid unnecessary calls to the apiserver
 * Enforce an interval in which version checks are performed to ensure that we catch apiserver ugprades


![screen shot 2019-03-05 at 3 09 14 pm](https://user-images.githubusercontent.com/1417913/53811257-37a96b80-3f59-11e9-87c9-8c58c149439e.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7292)
<!-- Reviewable:end -->

aanm commented on pkg/k8s/utils/factory.go:
```diff
@@ -45,8 +46,35 @@ var (
 	resourcers = map[reflect.Type]string{}
+
+	// lastEventReceived is the time at which the last k8s event was received
+	lastEventReceived eventTimestamper
```
Consider removing this change and adding this in the metric callback of k8s_watcher.go. My PR #7301 will remove this file altogether.

tgraf replied:
I've pushed the code into `pkg/k8s/metrics` and trigger the reset from all apiserver calls now so we can just remove the event handler when that code goes away.

tgraf committed:
k8s: Use apiserver interaction timestamp to determine status

Instead of relying entirely on version calls to the apiserver, consider the
timestamp of the last Kubernetes apisearver interaction to determine the
connectivity status.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
status: Add support for variable per-probe intervals

This allows to modify the interval based on whether the probe is stale or
failing and based on the current cluster size.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Record timestamp of last apiserver interaction

Introduce a package to track and provide the timestamp of the last apiserver
interaction.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Make k8s status check dependant on cluster size

The behavior so far has been to probe the k8s apiserver every 5 seconds. This
is very aggressive and puts a lot of load on the apiserver in very large
clusters. Make the check dependent on the cluster size.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Export ClusterSizeDependantInterval()

This allows other packages to use intervals which are dependent on the cluster
size.

Signed-off-by: Thomas Graf <thomas@cilium.io>

nebril commented on pkg/status/status.go:
```diff
@@ -57,6 +57,16 @@ type Probe struct {
 	// OnStatusUpdate is called whenever the status of the probe changes
 	OnStatusUpdate func(status Status)
+
+	// Interval always to specify a probe specific interval that can be
```
`s/always/allows` (I think?)

aanm replied:
if you read "always" fast enough it becomes "allows"

nebril commented on pkg/status/status_test.go:
```diff
@@ -47,6 +47,47 @@ func (s *StatusTestSuite) SetUpTest(c *C) {
+	// until we reach success
+	c.Assert(testutils.WaitUntil(func() bool {
+		return atomic.LoadUint64(&ok) >= 1
+	}, 1*time.Second), IsNil)
```
Can we lower this time to 10 or 50 milliseconds? 1s seems a bit excessive

tgraf replied:
This is the timeout for the test to fail, it returns earlier once the condition is met.
